nbcelltests/tests_vendored.py

```diff
@@ -46,10 +46,23 @@
 
 
 class TestNotebook(unittest.TestCase):
-    def setup_class(self):
-        self.kernel = RunningKernel("{kernel_name}")
 
-    def teardown_class(self):
+    KERNEL_NAME = "{kernel_name}"
+
+    @classmethod
+    def setUpClass(cls):
+        cls.kernel = RunningKernel(cls.KERNEL_NAME)
+
+    @classmethod
+    def tearDownClass(cls):
+        cls.kernel.stop()
```

Comment:
What is the class-level kernel used for?

Reply:
We ultimately want one kernel per class (i.e. one kernel per set of tests being run).

The hack we currently have to ensure not executing a cell multiple times in the same kernel is:

```
    # TODO: starting a new kernel per test is expensive, and
    # could be optimized.
    def setUp(self):
        self.kernel = RunningKernel(self.KERNEL_NAME)
```

So, starting a kernel in the `setUpClass` method is the one that will remain - starting a kernel in `setUp` will be removed.

Does that make sense? Or I am confused...?

Reply:
(Of course, you're right that at the moment the class level kernel is pointless, as it's immediately replaced by each test. But I'd rather keep the class level kernel for now, as it's what should be there if we weren't hacking things with a new kernel per test.)

Reply:
Ok, so we keep it in this diff as a forward looking change :)